### ISSUE TITLE
On synology "tailscale update" should maintain outbound connections configuration

### ISSUE THREAD

**Issue by yoyoma2** [tailscale/tailscale#12203]
### What is the issue?

Found on synology DSM 7.1

**Observed behavior:**
To enable outbound connections on synology, [tailscale says](https://tailscale.com/kb/1131/synology#enabling-synology-outbound-connections) to run a user-defined script at boot-up. This works but the "tailscale update" command installs and runs the new tailscale version but the outbound connections configuration is lost.

If my tailscale has auto-updated, rebooting the NAS or manually running that user defined boot-up task will allow Hyper-Backup to a remote NAS over tailscale to work again.

**Expected behavior:**
The "tailscale update" command detects if outbound connections are enabled and runs the tailscale configure-host script so the new version continues to support outbound connections.

**Useful command for diagnosing:**
This command shows the results of the tailscale configure-host script. It would be interesting to run it after a tailscale update command to see what has changed.
```
$ sudo getcap /var/packages/Tailscale/target/bin/tailscaled ; ls -l /dev/net/tun
/var/packages/Tailscale/target/bin/tailscaled = cap_net_admin,cap_net_raw+eip
crw-rw-rw- 1 root root 10, 200 May 13 09:03 /dev/net/tun
```
**How to test "tailscale update"?**
 Is it possible to test "tailscale update" when already up to date? Hard to test this with only monthly updates.

### Steps to reproduce

1) Use outbound connections on DSM 7
2) upgrade tailscale with "tailscale update"
Result: outbound connections now fail

### OS

Synology

### OS version

DSM 7.1

### Tailscale version

1.66.3

**Comment by yoyoma2** [comment]
Here's the diagnostic information before and after the update to tailscale 1.66.4. We see **the capabilities for the tailscale daemon are lost**. This is [expected behavior](https://unix.stackexchange.com/questions/283366/linux-file-capabilities-are-lost-when-i-modify-the-file-is-this-expected-behavi/283408#283408) so "tailscale update" should restore the capabilities on the new tailscaled file after updating it to maintain functionality.

```
$ tailscale --version ; sudo getcap /var/packages/Tailscale/target/bin/tailscaled ; ls -l /dev/net/tun
1.66.3
  tailscale commit: c231c3da2c07751077039f158e6836cafc8c5f63
  other commit: 7febb3078daab9ca04699c6a9750694d80942274
  go version: go1.22.3
/var/packages/Tailscale/target/bin/tailscaled = cap_net_admin,cap_net_raw+eip
crw-rw-rw- 1 root root 10, 200 May 13 09:03 /dev/net/tun
$
$ tailscale --version ; sudo getcap /var/packages/Tailscale/target/bin/tailscaled ; ls -l /dev/net/tun
1.66.4
  tailscale commit: 067defc641850791142cc83a44dadae29709344a
  other commit: 49481b7d0a74595aec90f60e4d9b784ed0036281
  go version: go1.22.3
crw-rw-rw- 1 root root 10, 200 May 13 09:03 /dev/net/tun
$
```

**Comment by awly** [comment]
Looks like you've diagnosed the issue correctly - because the `tailscaled` binary is replaced on upgrade, it loses the extra capabilities provisioned via https://tailscale.com/kb/1131/synology#enabling-synology-outbound-connections.

The immediate workaround is to re-run the script from the KB article, or simply to reboot the Synology to trigger it again. I'll update the KB article to mention that.

An idea for a more automated solution that @bradfitz suggested:
* create a wrapper binary that persists across upgrades
* `setcap` that wrapper binary instead of `tailscaled`
* the wrapper simply runs `tailscaled`
* on `tailscaled` startup, if we're missing the necessary capabilities and that wrapper binary exists, run it

**Comment by yoyoma2** [comment]
Until issue #12203 is resolved, [this auto-update script for tailscale on synology](https://www.synoforum.com/resources/tailscale-update-notifications-and-automatic-updates.206/updates) now includes a workaround in script version 2024.05.22.

**Comment by spacecakes** [comment]
@yoyoma2 has your hack worked reliably? 

This is an extremely annoying oversight that I wish would have been patched by now. I understand DSM is quirky and really is to blame, but since Tailscale is offering the service I wish it would receive some more attention

**Comment by yoyoma2** [comment]
> [@yoyoma2](https://github.com/yoyoma2) has your hack worked reliably?

Yes that auto-update script has been running reliably with fix_issue_12203 enabled for almost two years. I didn't get a notification for current v1.94.x so I investigated the script but it seems tailscale has skipped DSM this time and the spks are still at v1.92.3 as seen [here](https://pkgs.tailscale.com/stable/#spks). Hopfully v1.96.x gets a DSM build.

Agreed tailscale should have fixed this by now but for those who know about and use the auto-update script it's quickly forgotten.

**Comment by buntwo** [comment]
Just got bitten by this issue too, and spent an hour to fix it. Workaround for me is, like for others, to run the kb script after update.
Would love for Tailscale to fix this.

**Comment by yoyoma2** [comment]
The tailscale folks have incorrectly labeled this issue with "auto-update" which probably lowers the priority. The **manual update** with `tailscale update` is when the bug happens. The third-party auto-update script for synology DSM mentioned above includes a workaround for this bug.

All users updating tailscale normally/manually on DSM will encounter this bug if they use outbound connections. Tailscale should fix the **manual update** feature on synology DSM.
